repo/content/index_v2.go

```diff
@@ -0,0 +1,806 @@
+			return errors.Wrap(err, "unable to write entry")
+		}
+	}
+
+	// write pack ID entries in the index order of values from packID2Index (0, 1, 2, ...).
+	reversePackIDIndex := make([]blob.ID, len(b2.packID2Index))
+	for k, v := range b2.packID2Index {
+		reversePackIDIndex[v] = k
+	}
+
+	// emit pack ID information in this order.
+	for _, e := range reversePackIDIndex {
+		if err := b2.writePackIDEntry(w, e); err != nil {
+			return errors.Wrap(err, "error writing format info entry")
+		}
+	}
+
+	// build a list of indexV2FormatInfo using the order of indexes from uniqueFormatInfo2Index.
+	reverseFormatInfoIndex := make([]indexV2FormatInfo, len(b2.uniqueFormatInfo2Index))
+	for k, v := range b2.uniqueFormatInfo2Index {
+		reverseFormatInfoIndex[v] = k
+	}
+
+	// emit format information in this order.
+	for _, f := range reverseFormatInfoIndex {
+		if err := b2.writeFormatInfoEntry(w, f); err != nil {
+			return errors.Wrap(err, "error writing format info entry")
+		}
+	}
+
+	if _, err := w.Write(extraData); err != nil {
+		return errors.Wrap(err, "error writing extra data")
+	}
+
+	randomSuffix := make([]byte, v2RandomSuffixSize)
+	if _, err := rand.Read(randomSuffix); err != nil {
+		return errors.Wrap(err, "error getting random bytes for suffix")
+	}
+
+	if _, err := w.Write(randomSuffix); err != nil {
+		return errors.Wrap(err, "error writing extra random suffix to ensure indexes are always globally unique")
+	}
+
+	return w.Flush()
+}
+
+func (b *indexBuilderV2) prepareExtraData(sortedInfos []Info) []byte {
+	var extraData []byte
+
+	for _, it := range sortedInfos {
+		if it.GetPackBlobID() != "" {
+			if _, ok := b.packBlobIDOffsets[it.GetPackBlobID()]; !ok {
+				b.packBlobIDOffsets[it.GetPackBlobID()] = uint32(len(extraData))
+				extraData = append(extraData, []byte(it.GetPackBlobID())...)
+			}
+		}
+	}
+
+	b.extraDataOffset = v2IndexHeaderSize                                         // fixed header
+	b.extraDataOffset += uint32(b.entryCount * (b.keyLength + b.entrySize))       // entries index
+	b.extraDataOffset += uint32(len(b.packID2Index) * v2PackInfoSize)             // pack information
+	b.extraDataOffset += uint32(len(b.uniqueFormatInfo2Index) * v2FormatInfoSize) // formats
+
+	return extraData
+}
+
+func (b *indexBuilderV2) writeIndexEntry(w io.Writer, it Info) error {
+	var hashBuf [maxContentIDSize]byte
+
+	k := contentIDToBytes(hashBuf[:0], it.GetContentID())
+
+	if len(k) != b.keyLength {
+		return errors.Errorf("inconsistent key length: %v vs %v", len(k), b.keyLength)
+	}
+
+	if _, err := w.Write(k); err != nil {
+		return errors.Wrap(err, "error writing entry key")
+	}
+
+	if err := b.writeIndexValueEntry(w, it); err != nil {
+		return errors.Wrap(err, "error writing entry")
+	}
+
+	return nil
+}
+
+func (b *indexBuilderV2) writePackIDEntry(w io.Writer, packID blob.ID) error {
+	var buf [v2PackInfoSize]byte
+
+	buf[0] = byte(len(packID))
+	binary.BigEndian.PutUint32(buf[1:], b.packBlobIDOffsets[packID]+b.extraDataOffset)
```

Comment:
Should these offsets could be written relative to `extraDataOffset`, since they are relative to that anyway?